### PR TITLE
Kubectl user exec should accept zero-length environment values #652

### LINKED ISSUE #652: Kubectl user exec should accept zero-length environment values
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): Searched through current open issues

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.6-eks-d69f1b", GitCommit:"d69f1bf3669bf00b7f4a758e978e0e7a1e3a68f7", GitTreeState:"clean", BuildDate:"2019-02-28T20:26:10Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): MacOS
- **Kernel** (e.g. `uname -a`):
- **Install tools**: brew
- **Others**:

**What happened**:
```
Error in configuration: 
* env variable AWS_SECURITY_TOKEN value must be specified for eks-staging to use exec authentication plugin
* env variable AWS_SESSION_TOKEN value must be specified for eks-staging to use exec authentication plugin
* env variable AWS_SECRET_ACCESS_KEY value must be specified for eks-staging to use exec authentication plugin
* env variable AWS_ACCESS_KEY_ID value must be specified for eks-staging to use exec authentication plugin
```

**What you expected to happen**:
empty environment variable should be allowed

**How to reproduce it** (as minimally and precisely as possible):
Kubernetes config `users.[].user.exec.env.[].value = ""`

**Anything else we need to know**:
https://github.com/kubernetes/kubectl/blob/master/vendor/k8s.io/client-go/tools/clientcmd/validation.go#L253-L255 - these lines cause the problem. While it's totally reasonable to force a non-zero-length environment name (the lines immediately preceding, an empty environment value (overriding a value in the shell, for example) has utility and shouldn't be blocked. Asserting that `value` is present is reasonable, asserting that it's not empty is not.

### PR DESCRIPTION
**Pull request by pswica**
**What type of PR is this?**
/kind feature
**What this PR does / why we need it**:
Allow blank strings "" as values in `users.[].user.exec.env.[].value`
**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubectl/issues/652
**Special notes for your reviewer**:
**Does this PR introduce a user-facing change?**:
```release-note
Users can now enter blank strings as values in the `users.[].user.exec.env.[].value` section of a kubeconfig file.
```

### CODE REVIEW

**Comment by pswica** [comment]
@kubernetes/sig-cli-pr-reviews PTAL

File: staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go [hunk 1]
```diff
@@ -385,79 +385,76 @@ func TestValidateAuthInfoExec(t *testing.T) {
 	test.testConfig(t)
 }
```
Comment (asalkeld) [inline staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go]: the moving functions around makes it more difficult than needed to review.

**Review by asalkeld — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go]: agree... please put these back in the previous order so the diff reflects what is actually being changed

File: staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go [hunk 2]
```diff
@@ -443,22 +443,19 @@ func TestValidateAuthInfoExecWithAuthProvider(t *testing.T) {
...
 
-func TestValidateAuthInfoExecInvalidEnv(t *testing.T) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/validation_test.go]: can you keep this test case but make the invalid aspect a zero-length envvar name? a new testcase to ensure zero-length values are accepted is fine

**Comment by pswica** [comment]
Thank you @liggitt @asalkeld I reordered the tests

**Pull request merged**